lib/errors.ts

```diff
@@ -14,18 +14,76 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-import errors = require('resin-cli-errors');
-import patterns = require('./utils/patterns');
+import _ = require('lodash');
+import os = require('os');
 import Raven = require('raven');
 import Promise = require('bluebird');
+import { stripIndent } from 'common-tags';
+
+import patterns = require('./utils/patterns');
 
 const captureException = Promise.promisify<string, Error>(
 	Raven.captureException,
 	{ context: Raven },
 );
 
+function hasCode(error: any): error is Error & { code: string } {
```

Comment:
Shouldn't the return type be `boolean`?
I don't really understand this return type.

Reply:
This is a [user-defined type guard](http://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).

It's basically a more specific type of boolean: not only does it return a boolean value, the value also gives TypeScript some information about types, which it can use infer the type of the variable in case that uses this test.

As a specific example, in this case this means that inside the `if (hasCode(error)) { ... }` block, we're allowed to look at `error.code`, because TS can infer that this method has returned true, and this type signatures says that this means the object has a `.code` property. If this just returns boolean, that won't work, because it's just an `Error`, and `.code` isn't a property that all errors have.